File: Dynamic_Programming/131.Palindrome-Partitioning/131.Palindrome-Partitioning.cpp
class Solution {
    int dp[16][16];
    vector<vector<string>>rets;
    int n;
    string s;
public:
    vector<vector<string>> partition(string s) 
    {
        this->s = s;
        n = s.size();
        
        for (int i=0; i<n; i++)
            dp[i][i] = 1;
        for (int i=0; i+1<n; i++)
            dp[i][i+1] = s[i]==s[i+1];
        for (int len = 3; len <=n; len++)
            for (int i=0; i+len-1<n; i++)
            {
                int j = i+len-1;
                if (s[i]==s[j])
                    dp[i][j] = dp[i+1][j-1];
                else
                    dp[i][j] = 0;
            }

        vector<string>temp;
        dfs(0, temp);

        return rets;
    }

    void dfs(int i, vector<string>&temp)
    {
        if (i==n)
        {
            rets.push_back(temp);
            return;
        } 
        for (int j=i; j<n; j++)
        {
            if (dp[i][j])
            {
                temp.push_back(s.substr(i,j-i+1));
                dfs(j+1, temp);
                temp.pop_back();
            }
        }
    }
};

File: Dynamic_Programming/131.Palindrome-Partitioning/131.Palindrome-Partitioning.py
class Solution:
    def partition(self, s):
        """
        :type s: str
        :rtype: List[List[str]]
        """
        N = len(s)
        dp = [[0 for _ in range(N)] for _ in range(N)]
        for i in range(N):
            dp[i][i]=1
        for i in range(N-1):
            if s[i]==s[i+1]: dp[i][i+1]=1
        for length in range(3,N+1):
            for start in range(N-length+1):
                if s[start]==s[start+length-1] and dp[start+1][start+length-2]==1:
                    dp[start][start+length-1]=1     
                                   
        ret,combo = [],[]
        def DFS(start,combo,ret):
            if start==N: 
                ret.append(combo)
                return
            for i in range(start,N):
                if dp[start][i]==1: 
                    DFS(i+1,combo+[s[start:i+1]],ret)
            return        
        
        DFS(0,combo,ret)
        return ret
